src/main/java/com/slack/kaldb/metadata/ZookeeperMetadataStore.java

```diff
@@ -0,0 +1,337 @@
+
+import static com.slack.kaldb.util.ArgValidationUtils.ensureNonEmptyString;
+import static com.slack.kaldb.util.ArgValidationUtils.ensureTrue;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.slack.kaldb.util.FatalErrorHandler;
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.MeterRegistry;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import org.apache.curator.RetryPolicy;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.CuratorFrameworkFactory;
+import org.apache.curator.framework.api.CuratorEventType;
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.Watcher;
+import org.apache.zookeeper.data.Stat;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * ZookeeperMetadataStore implements a store in zookeeper. This class interacts with ZK using the
+ * apache curator library. It also abstracts it's users from the details of ZK and provides a
+ * simpler higher level API.
+ *
+ * <p>All the public calls in this class are executed in it's own thread pool which makes the API
+ * non-blocking while allowing for parallelism. The actual methods are implemented as private
+ * methods ending in Impl. A corresponding public method is exposed that executes these calls in a
+ * thread pool.
+ *
+ * <p>NOTE: We will use the async as a suffix for methods which run in the background in curator.
+ * So, we don't use it for calls that run in a future pool even though they are technically async.
+ */
+public class ZookeeperMetadataStore implements MetadataStore {
+  private static final Logger LOG = LoggerFactory.getLogger(ZookeeperMetadataStore.class);
+
+  public static final String METADATA_FAILED_COUNTER = "metadata.failed";
+  public static final String ZK_FAILED_COUNTER = "metadata.failed.zk";
+  public static final String METADATA_WRITE_COUNTER = "metadata.write";
+  public static final String METADATA_READ_COUNTER = "metadata.read";
+
+  private final CuratorFramework curator;
+  private final Counter failureCounter;
+  private final Counter zkFailureCounter;
+  private final Counter metadataWriteCounter;
+  private final Counter metadataReadCounter;
+
+  // A thread pool to run all the metadata store operations in.
+  private final ListeningExecutorService metadataExecutorService;
+
+  public ZookeeperMetadataStore(
+      String zkHostPath,
+      String zkPathPrefix,
+      int sessionTimeoutMs,
+      int connectionTimeoutMs,
+      RetryPolicy retryPolicy,
```

Comment:
we _might_ need a retry policy to be configurable based on the operation - I know for solr there were some operations that were best effort , some that could not be retried etc. I'll have to refresh my memory but thought I'd point out.

for now I think it's okay to keep it simple till we feel the need to

Reply:
Agree. Yeah, this is basic API for ZK CRUD. We will have another layer on top (one for snapshot metadata and one for node metadata etc..) on top of this API which will configure the setting for each use case. All those customizations would go into that API. The next diff would clarify this configuration more.